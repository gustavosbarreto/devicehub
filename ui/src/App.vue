<template>
<v-app>
    <v-navigation-drawer v-if="isLoggedIn" :clipped="false" fixed v-model="drawer" :mini-variant="false" enable-resize-watcher app>
        <v-container>
            <div class="text-center">
                <v-icon>mdi-console</v-icon>
                <h2 style="font-family: monospace">ShellHub</h2>
                <span class="overline">beta</span>
            </div>
        </v-container>
        <v-list>
            <v-list-item v-for="item in items" :key="item.title" :to="item.path" two-line>
                <v-list-item-action>
                    <v-icon v-text="item.icon"></v-icon>
                </v-list-item-action>

                <v-list-item-content>
                    <v-list-item-title v-text="item.title"></v-list-item-title>
                </v-list-item-content>
            </v-list-item>
        </v-list>
    </v-navigation-drawer>
    <v-app-bar app elevate-on-scroll color="primary" class="pl-3 pr-4" v-if="isLoggedIn">
        <v-menu transition="scale-transition" origin="top left">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-server</v-icon>
                    My Device Fleet
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>
        </v-menu>
        <v-spacer></v-spacer>
        <v-chip>
            <v-icon>help</v-icon>
        </v-chip>
        <v-chip>
            <v-icon>notifications</v-icon>
        </v-chip>
        <v-menu transition="scale-transition" origin="top right">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-account</v-icon>
                    {{ $store.getters["auth/currentUser"] }}
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>

            <v-card>
                <v-list-item three-line>
                    <v-list-item-content>
                        <v-list-item-title class="mb-1">Tenant ID</v-list-item-title>
                        <v-list-item-subtitle>
                            <v-chip>
                                <span>
                                    {{ tenant }}
                                </span>
                                <v-icon right v-clipboard="tenant" v-clipboard:success="() => { copySnack = true; }">mdi-content-copy</v-icon>
                            </v-chip>
                        </v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>

                <v-card-actions>
                    <v-btn small text @click="logout()">
                        Logout
                    </v-btn>
                </v-card-actions>
            </v-card>
        </v-menu>
    </v-app-bar>
    <v-content>
        <v-container class="pa-8" fluid>
            <router-view></router-view>
        </v-container>
        <v-snackbar v-model="copySnack" :timeout=3000>Tenant ID copied to clipboard</v-snackbar>
    </v-content>
</v-app>
</template>

<script>
export default {
  name: "App",

  methods: {
    logout() {
      this.$store.dispatch("auth/logout").then(() => {
        this.$router.push("/login");
      });
    }
  },

  computed: {
    tenant() {
      return this.$store.getters["auth/tenant"];
    },

    isLoggedIn() {
      return this.$store.getters["auth/isLoggedIn"];
    }
  },

  data() {
    return {
      drawer: true,
      clipped: false,
      copySnack: false,
      items: [
        {
          icon: "dashboard",
          title: "Dashboard",
          path: "/"
        },
        {
          icon: "devices",
          title: "Devices",
          path: "/devices"
        },
        {
          icon: "history",
          title: "Sessions",
          path: "/sessions"
        }
      ]
    };
  }
};
</script>

<style>
.v-list-active {
  border-left: 4px solid var(--v-primary-base);
}

.text-shadow {
  text-shadow: #000 0 0 6px;
  color: transparent;
}
</style>
