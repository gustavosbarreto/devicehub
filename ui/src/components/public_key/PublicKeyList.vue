<template>
  <fragment>
    <v-card-text class="pa-0">
      <v-data-table
        :headers="headers"
        :items="getPublicKeys"
        item-key="fingerprint"
        :sort-by="['started_at']"
        :sort-desc="[true]"
        :items-per-page="10"
        :footer-props="{'items-per-page-options': [10, 25, 50, 100]}"
        :server-items-length="getNumberPublicKeys"
        :options.sync="pagination"
        :disable-sort="true"
        data-test="publicKeyList-dataTable"
      >
        <template #[`item.name`]="{ item }">
          {{ item.name }}
        </template>

        <template #[`item.fingerprint`]="{ item }">
          {{ item.fingerprint }}
        </template>

        <template #[`item.filter`]="{ item }">
          <div
            v-if="filterKey(item.filter)=='hostname'"
          >
            {{ formatHostnameFilter(item.filter) }}
          </div>

          <div v-else-if="filterKey(item.filter)=='tags'">
            <v-tooltip
              v-for="(tag, index) in item.filter.tags"
              :key="index"
              bottom
              :disabled="!showTag(tag)"
            >
              <template #activator="{ on, attrs }">
                <v-chip
                  class="ml-1 mb-1"
                  small
                  outlined
                  v-bind="attrs"
                  v-on="on"
                >
                  {{ displayOnlyTenCharacters(tag) }}
                </v-chip>
              </template>

              <span v-if="showTag(tag)">
                {{ tag }}
              </span>
            </v-tooltip>
          </div>
        </template>

        <template #[`item.username`]="{ item }">
          {{ item.username === '' ? 'All users' : item.username }}
        </template>

        <template #[`item.created_at`]="{ item }">
          {{ item.created_at | moment("ddd, MMM Do YY, h:mm:ss a") }}
        </template>

        <template #[`item.actions`]="{ item }">
          <v-menu
            :ref="'menu'+getPublicKeys.indexOf(item)"
            offset-y
          >
            <template #activator="{ on, attrs }">
              <v-chip
                color="transparent"
                v-on="on"
              >
                <v-icon
                  small
                  class="icons"
                  v-bind="attrs"
                  v-on="on"
                >
                  mdi-dots-horizontal
                </v-icon>
              </v-chip>
            </template>

            <v-card>
              <v-tooltip
                bottom
                :disabled="hasAuthorizationFormDialogEdit"
              >
                <template #activator="{ on, attrs }">
                  <div
                    v-bind="attrs"
                    v-on="on"
                  >
                    <v-list-item
                      :disabled="!hasAuthorizationFormDialogEdit"
                      @click="showPublicKeyFormDialog(getPublicKeys.indexOf(item))"
                    >
                      <PublicKeyFormDialogEdit
                        :key-object="item"
                        :show.sync="publicKeyFormDialogShow[getPublicKeys.indexOf(item)]"
                        @update="refresh"
                      />
                    </v-list-item>
                  </div>
                </template>

                <span>
                  You don't have this kind of authorization.
                </span>
              </v-tooltip>

              <v-tooltip
                bottom
                :disabled="hasAuthorizationFormDialogRemove"
              >
                <template #activator="{ on, attrs }">
                  <div
                    v-bind="attrs"
                    v-on="on"
                  >
                    <v-list-item
                      :disabled="!hasAuthorizationFormDialogRemove"
                      @click="showPublicKeyDelete(getPublicKeys.indexOf(item))"
                    >
                      <PublicKeyDelete
                        :fingerprint="item.fingerprint"
                        :show.sync="publicKeyDeleteShow[getPublicKeys.indexOf(item)]"
                        @update="refresh"
                      />
                    </v-list-item>
                  </div>
                </template>

                <span>
                  You don't have this kind of authorization.
                </span>
              </v-tooltip>
            </v-card>
          </v-menu>
        </template>
      </v-data-table>
    </v-card-text>
  </fragment>
</template>

<script>

import PublicKeyFormDialogEdit from '@/components/public_key/PublicKeyFormDialogEdit';
import PublicKeyDelete from '@/components/public_key/PublicKeyDelete';

import hasPermission from '@/components/filter/permission';

export default {
  name: 'PublicKeyListComponent',

  filters: { hasPermission },

  components: {
    PublicKeyFormDialogEdit,
    PublicKeyDelete,
  },

  data() {
    return {
      pagination: {},
      publicKeyFormDialogShow: [],
      publicKeyDeleteShow: [],
      editAction: 'edit',
      removeAction: 'remove',
      headers: [
        {
          text: 'Name',
          value: 'name',
          align: 'center',
        },
        {
          text: 'Fingerprint',
          value: 'fingerprint',
          align: 'center',
        },
        {
          text: 'Filter',
          value: 'filter',
          align: 'center',
        },
        {
          text: 'Username',
          value: 'username',
          align: 'center',
        },
        {
          text: 'Created At',
          value: 'created_at',
          align: 'center',
        },
        {
          text: 'Actions',
          value: 'actions',
          align: 'center',
        },
      ],
    };
  },

  computed: {
    getPublicKeys() {
      return this.$store.getters['publickeys/list'];
    },

    getNumberPublicKeys() {
      return this.$store.getters['publickeys/getNumberPublicKeys'];
    },

    hasAuthorizationFormDialogEdit() {
      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.publicKey[this.editAction],
        );
      }

      return false;
    },

    hasAuthorizationFormDialogRemove() {
      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.publicKey[this.removeAction],
        );
      }

      return false;
    },
  },

  watch: {
    pagination: {
      handler() {
        this.getPublicKeysList();
      },
      deep: true,
    },
  },

  methods: {
    showTag(str) {
      if (str !== undefined) {
        if (str.length > 10) {
          return true;
        }
      }
      return false;
    },

    filterKey(filter) {
      return Reflect.ownKeys(filter)[0];
    },

    hasKeys(filter) {
      return Object.keys(filter).length > 0;
    },

    displayOnlyTenCharacters(str) {
      if (str !== undefined) {
        if (str.length > 10) return `${str.substr(0, 10)}...`;
      }
      return str;
    },

    refresh() {
      this.getPublicKeysList();
    },

    async getPublicKeysList() {
      if (!this.$store.getters['boxs/getStatus']) {
        const data = {
          perPage: this.pagination.itemsPerPage,
          page: this.pagination.page,
        };

        try {
          await this.$store.dispatch('publickeys/fetch', data);
        } catch {
          this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.publicKeyList);
        }
      } else {
        this.setArrays();
        this.$store.dispatch('boxs/setStatus', false);
      }
    },

    showPublicKeyFormDialog(index) {
      this.publicKeyFormDialogShow[index] = this.publicKeyFormDialogShow[index] === undefined
        ? true : !this.publicKeyFormDialogShow[index];
      this.$set(this.publicKeyFormDialogShow, index, this.publicKeyFormDialogShow[index]);

      this.closeMenu(index);
    },

    showPublicKeyDelete(index) {
      this.publicKeyDeleteShow[index] = this.publicKeyDeleteShow[index] === undefined
        ? true : !this.publicKeyDeleteShow[index];
      this.$set(this.publicKeyDeleteShow, index, this.publicKeyDeleteShow[index]);

      this.closeMenu(index);
    },

    formatHostnameFilter(filter) {
      return filter.hostname === '.*' ? 'All devices' : filter.hostname;
    },

    setArrays() {
      const numberPublicKey = this.getPublicKeys.length;

      if (numberPublicKey > 0) {
        this.publicKeyFormDialogShow = new Array(numberPublicKey).fill(false);
        this.publicKeyDeleteShow = new Array(numberPublicKey).fill(false);
      }
    },

    closeMenu(index) {
      this.$refs[`menu${index}`].isActive = false;
    },
  },
};

</script>
