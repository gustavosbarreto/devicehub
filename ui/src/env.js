export default {
  /* eslint-disable no-param-reassign */
  install(Vue) {
    const env = {
      isHosted: window.env.HOSTED === 'true',
    };

    Vue.env = env;
    Vue.prototype.$env = env;
  },
  /* eslint-enable no-param-reassign */
};
