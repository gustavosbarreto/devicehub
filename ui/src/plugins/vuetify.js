import Vue from 'vue';
import Vuetify from 'vuetify/lib';
import '@mdi/font/css/materialdesignicons.css';
import 'font-logos/assets/font-logos.css';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { fas } from '@fortawesome/free-solid-svg-icons';
import Clipboard from 'v-clipboard';
import Fragment from 'vue-fragment';

//import 'vuetify/dist/vuetify.min.css'
//import '../styles/variables.scss'

Vue.component('font-awesome-icon', FontAwesomeIcon); // Register component globally
library.add(fas); // Include needed icons.

Vue.use(Vuetify);
Vue.use(Clipboard);
Vue.use(Fragment.Plugin);

const DEVICE_ICONS = {
  alpine: 'fl-alpine',
  arch: 'fl-archlinux',
  centos: 'fl-centos',
  coreos: 'fl-coreos',
  debian: 'fl-debian',
  devuan: 'fl-devuan',
  elementary: 'fl-elementary',
  fedora: 'fl-fedora',
  freebsd: 'fl-freebsd',
  gentoo: 'fl-gentoo',
  linuxmint: 'fl-linuxmint',
  mageia: 'fl-mageia',
  manjaro: 'fl-manjaro',
  mandriva: 'fl-mandriva',
  nixos: 'fl-nixos',
  opensuse: 'fl-opensuse',
  rhel: 'fl-redhat',
  sabayon: 'fl-sabayon',
  slackware: 'fl-slackware',
  ubuntu: 'fl-ubuntu',
  raspbian: 'fl-raspberry-pi',
  'ubuntu-core': 'fl-ubuntu',
  void: 'fl-void',
};

export default new Vuetify({
  iconfont: 'md',
  icons:{
    values:DEVICE_ICONS
  },
  theme: {
    dark: false,
    themes: {
      light: {
        primary: '#364576',
        secondary: '#b0bec5',
        accent: '#8c9eff',
        error: '#b71c1c',
      },
    }
  }
});
