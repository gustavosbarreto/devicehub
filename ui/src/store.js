import Vue from 'vue';
import Vuex from 'vuex';
import stats from '@/modules/stats';
import sessions from '@/modules/sessions';
import auth from '@/modules/auth';
import devices from '@/modules/devices';
import modals from '@/modules/modals';
import firewallrules from '@/modules/firewall_rules';
import notifications from '@/modules/notifications';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    devices, modals, stats, sessions, auth, firewallrules, notifications,
  },
});
