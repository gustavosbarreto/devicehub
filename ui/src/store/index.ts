import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";

import { auth, AuthState } from "./modules/auth";
import { apiKeys, ApiKeysState } from "./modules/api_keys";
import { layout, LayoutState } from "./modules/layout";
import { users, UsersState } from "./modules/users";
import { tags, TagsState } from "./modules/tags";
import { stats, StatsState } from "./modules/stats";
import { spinner, SpinnerState } from "./modules/spinner";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { sessions, SessionsState } from "./modules/sessions";
import { security, SecurityState } from "./modules/security";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { privateKey, PrivateKeyState } from "./modules/private_key";
import { notifications, NotificationsState } from "./modules/notifications";
import { modal, ModalState } from "./modules/modal";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { devices, DevicesState } from "./modules/devices";
import { box, BoxState } from "./modules/box";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { billing } from "./modules/billing";
import { customer, CustomerState } from "./modules/customer";
import { announcement, AnnouncementState } from "./modules/announcement";
import apiPlugin from "./plugins/api";

export interface State {
  auth: AuthState;
  apiKeys: ApiKeysState;
  billing: NamespacesState;
  customer: CustomerState;
  box: BoxState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  modal: ModalState;
  namespaces: NamespacesState;
  notifications: NotificationsState;
  privateKey: PrivateKeyState;
  publicKeys: PublicKeysState;
  security: SecurityState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  tags: TagsState;
  users: UsersState;
  announcement: AnnouncementState;
}

export const key: InjectionKey<Store<State>> = Symbol("store");

export const store = createStore<State>({
  modules: {
    auth,
    apiKeys,
    customer,
    billing,
    box,
    devices,
    firewallRules,
    layout,
    modal,
    namespaces,
    notifications,
    privateKey,
    publicKeys,
    security,
    sessions,
    snackbar,
    spinner,
    stats,
    tags,
    users,
    announcement,
  },
  plugins: [
    apiPlugin,
  ],
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
