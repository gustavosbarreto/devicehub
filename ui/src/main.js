import Vue from 'vue';

import Snackbar from '@/components/snackbar/Snackbar';

import App from './App';
import router from './router';
import store from './store';
import env from './env';
import './vee-validate';
import vuetify from './plugins/vuetify';

Vue.config.productionTip = false;

Vue.component('Snackbar', Snackbar);

Vue.use(require('vue-moment'));

Vue.use(env);

new Vue({
  vuetify,
  router,
  store,
  render: (h) => h(App),
}).$mount('#app');
