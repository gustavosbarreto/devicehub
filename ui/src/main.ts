import { createApp } from "vue";
import * as Sentry from "@sentry/vue";
import { BrowserTracing } from "@sentry/tracing";
import { envVariables } from "./envVariables";
import vuetify from "./plugins/vuetify";
import { key, store } from "./store";
import { router } from "./router";
import App from "./App.vue";

import { loadFonts } from "./plugins/webfontloader";

import SnackbarComponent from "./components/Snackbar/Snackbar.vue";
import { SnackbarPlugin } from "./plugins/snackbar";

const app = createApp(App);

Sentry.init({
  app,
  dsn: envVariables.sentryDsn || "",
  integrations: [
    new BrowserTracing({
      routingInstrumentation: Sentry.vueRouterInstrumentation(router),
    }),
  ],
  tracesSampleRate: 1.0,
  hooks: ["activate", "create", "destroy", "mount", "update"],
  timeout: 500,
  release: envVariables.version || "latest",
});
Sentry.setTag("project", "shellhub-ui");

loadFonts();

app.use(vuetify);
app.use(router);
app.use(store, key);
app.use(SnackbarPlugin);
app.component("SnackbarComponent", SnackbarComponent);
app.mount("#app");
