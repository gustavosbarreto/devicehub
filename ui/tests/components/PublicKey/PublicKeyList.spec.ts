import { createVuetify } from "vuetify";
import { mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";
import PublicKeyList from "../../../src/components/PublicKeys/PublicKeysList.vue";
import { createStore } from "vuex";
import { key } from "../../../src/store";
import routes from "../../../src/router";

const numberPublicKeysGlobal = 2;

const publicKeysGlobal = [
  {
    data: "BBGVvbmF",
    fingerprint: "00:00:00",
    created_at: "2020-11-23T20:59:13.323Z",
    tenant_id: "xxxxxxxx",
    name: "shellhub",
  },
  {
    data: "AbGVvbmF",
    fingerprint: "00:00:00",
    created_at: "2020-11-23T20:59:13.323Z",
    tenant_id: "xxxxxxxx",
    name: "shellhub",
  },
];

const headers = [
  {
    text: "Name",
    value: "name",
  },
  {
    text: "Fingerprint",
    value: "fingerprint",
  },
  {
    text: "Filter",
    value: "filter",
  },
  {
    text: "Username",
    value: "username",
  },
  {
    text: "Created At",
    value: "created_at",
  },
  {
    text: "Actions",
    value: "actions",
  },
];

const tests = [
  {
    description: "List data when user has owner role",
    role: {
      type: "owner",
      permission: true,
    },
    variables: {
      publicKeysGlobal,
      numberPublicKeysGlobal,
    },
    data: {
      pagination: {},
      loading: false,
      itemsPerPage: 10,
      page: 1,
      publicKeyFormDialogShow: [],
      publicKeyDeleteShow: [],
      editAction: "edit",
      headers,
    },
    computed: {
      publicKeys: publicKeysGlobal,
      getNumberPublicKeys: numberPublicKeysGlobal,
      hasAuthorizationFormDialogEdit: true,
      hasAuthorizationFormDialogRemove: true,
    },
  },
  {
    description: "List data when user has operator role",
    role: {
      type: "operator",
      permission: false,
    },
    variables: {
      publicKeysGlobal,
      numberPublicKeysGlobal,
    },
    data: {
      pagination: {},
      loading: false,
      itemsPerPage: 10,
      page: 1,
      publicKeyFormDialogShow: [],
      publicKeyDeleteShow: [],
      editAction: "edit",
      headers,
    },
    computed: {
      publicKeys: publicKeysGlobal,
      getNumberPublicKeys: numberPublicKeysGlobal,
      hasAuthorizationFormDialogEdit: false,
      hasAuthorizationFormDialogRemove: false,
    },
  },
];

const store = (
  publicKeys: any,
  numberPublicKeys: any,
  status: any,
  currentRole: any
) => {
  return createStore({
    state: {
      publicKeys,
      numberPublicKeys,
      status,
      currentRole,
    },
    getters: {
      "publicKeys/list": (state) => state.publicKeys,
      "publicKeys/getNumberPublicKeys": (state) => state.numberPublicKeys,
      "box/getStatus": (state) => state.status,
      "auth/role": (state) => state.currentRole,
    },
    actions: {
      "publicKeys/fetch": vi.fn(),
      "snackbar/showSnackbarErrorLoading": vi.fn(),
      "box/setStatus": vi.fn(),
    },
  });
};

describe("PublicKeyList", () => {
  let wrapper: VueWrapper<any>;
  const vuetify = createVuetify();

  tests.forEach((test) => {
    describe(`${test.description}`, () => {
      beforeEach(() => {
        wrapper = mount(PublicKeyList, {
          global: {
            plugins: [
              [
                store(
                  test.variables.publicKeysGlobal,
                  test.variables.numberPublicKeysGlobal,
                  test.role.type,
                  test.role.permission
                ),
                key,
              ],
              routes,
              vuetify,
            ],
          },
          shallow: true,
        });
      });

      ///////
      // Component Rendering
      //////

      it("Is a Vue instance", () => {
        expect(wrapper).toBeTruthy();
      });
      it("Renders the component", () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      ///////
      // Data and Props checking
      //////

      it("Data is defined", () => {
        expect(wrapper.vm.$data).toBeDefined();
      });

      it("Compare data with default value", () => {
        expect(wrapper.vm.headers).toStrictEqual(headers);
        expect(wrapper.vm.loading).toStrictEqual(test.data.loading);
        expect(wrapper.vm.page).toStrictEqual(test.data.page);
        expect(wrapper.vm.itemsPerPage).toStrictEqual(test.data.itemsPerPage);
      });

      it("Compare the computed with the default value", () => {
        expect(wrapper.vm.publicKeys).toStrictEqual(test.computed.publicKeys);
        expect(wrapper.vm.getNumberPublicKeys).toStrictEqual(
          test.computed.getNumberPublicKeys
        );
      });

      //////
      // HTML validation
      //////

      it('Renders the template with data', () => {
        const dt = wrapper.find('[data-test="publicKeys-list"]');
        expect(dt.exists()).toBeTruthy();
      });
    });
  });
});
