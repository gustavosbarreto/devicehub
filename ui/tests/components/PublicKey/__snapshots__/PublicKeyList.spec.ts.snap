// Vitest Snapshot v1

exports[`PublicKeyList > List data when user has operator role > Renders the component 1`] = `
"<div>
  <data-table-stub itemsperpage=\\"10\\" totalcount=\\"2\\" headers=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" items=\\"[object Object],[object Object]\\" comboboxoptions=\\"10,20,50,100\\" loading=\\"false\\" actualpage=\\"1\\" nextpage=\\"async () => {
      await getPublicKeysList(itemsPerPage.value, ++page.value);
    }\\" previouspage=\\"async () => {
      try {
        if (page.value > 1)
          await getPublicKeysList(itemsPerPage.value, --page.value);
      } catch (error) {
        store.dispatch(&quot;snackbar/setSnackbarErrorDefault&quot;);
        throw new Error(error);
      }
    }\\" data-test=\\"publicKeys-list\\"></data-table-stub>
</div>"
`;

exports[`PublicKeyList > List data when user has owner role > Renders the component 1`] = `
"<div>
  <data-table-stub itemsperpage=\\"10\\" totalcount=\\"2\\" headers=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" items=\\"[object Object],[object Object]\\" comboboxoptions=\\"10,20,50,100\\" loading=\\"false\\" actualpage=\\"1\\" nextpage=\\"async () => {
      await getPublicKeysList(itemsPerPage.value, ++page.value);
    }\\" previouspage=\\"async () => {
      try {
        if (page.value > 1)
          await getPublicKeysList(itemsPerPage.value, --page.value);
      } catch (error) {
        store.dispatch(&quot;snackbar/setSnackbarErrorDefault&quot;);
        throw new Error(error);
      }
    }\\" data-test=\\"publicKeys-list\\"></data-table-stub>
</div>"
`;
